cmd/agent/subcommands/snmp/command.go

```diff
@@ -413,61 +420,18 @@ func scanDevice(connParams *connectionParams, args argsType, conf config.Compone
 	if err := snmp.Connect(); err != nil {
 		return fmt.Errorf("unable to connect to SNMP agent on %s:%d: %w", snmp.LocalAddr, snmp.Port, err)
 	}
-	defer snmp.Conn.Close()
-	pdus, err := gatherPDUs(snmp)
-	if err != nil {
-		return err
-	}
 
 	namespace := conf.GetString("network_devices.namespace")
-	deviceID := namespace + ":" + deviceAddr
-	var deviceOids []*metadata.DeviceOID
-	for _, pdu := range pdus {
-		record, err := metadata.DeviceOIDFromPDU(deviceID, pdu)
-		if err != nil {
-			logger.Warnf("PDU parsing error: %v", err)
-			continue
-		}
-		deviceOids = append(deviceOids, record)
-	}
-	forwarder, err := demux.GetEventPlatformForwarder()
-	if err != nil {
-		return fmt.Errorf("unable to get sender: %w", err)
-	}
-	metadataPayloads := metadata.BatchDeviceScan(namespace, time.Now(), metadata.PayloadMetadataBatchSize, deviceOids)
-	for _, payload := range metadataPayloads {
-		payloadBytes, err := json.Marshal(payload)
-		if err != nil {
-			logger.Errorf("Error marshalling device metadata: %v", err)
-			continue
-		}
-		m := message.NewMessage(payloadBytes, nil, "", 0)
-		logger.Debugf("Device OID metadata payload is %d bytes", len(payloadBytes))
-		logger.Tracef("Device OID metadata payload: %s", string(payloadBytes))
-		if err := forwarder.SendEventPlatformEventBlocking(m, eventplatform.EventTypeNetworkDevicesMetadata); err != nil {
-			return err
-		}
-	}
 
-	return nil
-}
-
-// gatherPDUs returns PDUs from the given SNMP device that should cover ever
-// scalar value and at least one row of every table.
-func gatherPDUs(snmp *gosnmp.GoSNMP) ([]*gosnmp.SnmpPDU, error) {
-	var pdus []*gosnmp.SnmpPDU
-	err := gosnmplib.ConditionalWalk(snmp, "", false, func(dataUnit gosnmp.SnmpPDU) (string, error) {
-		pdus = append(pdus, &dataUnit)
-		return gosnmplib.SkipOIDRowsNaive(dataUnit.Name), nil
-	})
+	err = snmpScanner.RunDeviceScan(snmp, namespace)
 	if err != nil {
-		return nil, err
+		fmt.Fprintf(os.Stderr, "Unable to perform device scan: %v\n", err)
```

Comment:
How are you deciding which errors should be returned and which should just be printed? It feels like since scanDevice can return an error it should do so instead of printing to stderr and returning nil.

Reply:
This one is a mistake, and it should return an error indeed

Reply:
I've kept the logic similar for the rest, the only place where we log the error is when we cannot contact the Agent because this is recoverable error (if there are enough arguments)